epinorm/norm.py

```diff
@@ -301,38 +346,70 @@ def _add_source_details(self):
 
     def _geocode(self):
         """Geocode EMPRES-i data."""
-        locality_names = []
-        locality_ids = []
-        admin_level_1_names = []
-        admin_level_1_ids = []
-        country_names = []
-        country_ids = []
-        for index, row in self._data.iterrows():
+
+        country_name_to_code = pd.read_csv(COUNTRIES_FILE, index_col="name")["alpha_2"].to_dict()
+        country_name_to_code.update(COUNTRIES_EXCEPTIONS) # add the exceptions
+
+        admin_level_1_table = pd.read_csv(ADMIN_LEVEL_1_FILE)
+        admin_level_1_table = admin_level_1_table.replace(np.nan, None) # so its easier to deal with those values
+
+        # output columns
+        locality_names = np.full(len(self._data), None)
+        locality_ids = np.full(len(self._data), None)
+        admin_level_1_names = np.full(len(self._data), None)
+        admin_level_1_ids = np.full(len(self._data), None)
+        country_names = np.full(len(self._data), None)
+        country_ids = np.full(len(self._data), None)
```

Comment:
Add a new method get_default_array() to the DataHandler class that encapsulates the repetitive logic: np.full(len(self._data), None) and use it instead.